vispy/visuals/filters/mesh.py

```diff
@@ -105,3 +107,232 @@ def _update_texcoords_buffer(self, texcoords):
+        self._enabled = enabled
+        self._update_data()
+
+    @property
+    def light_dir(self):
+        """The light direction."""
+        return self._light_dir
+
+    @light_dir.setter
+    def light_dir(self, direction):
+        direction = np.array(direction, float).ravel()
+        if direction.size != 3 or not np.isfinite(direction).all():
+            raise ValueError('Invalid direction %s' % direction)
+        self._light_dir = tuple(direction)
+        self._update_data()
+
+    @property
+    def light_color(self):
+        """The light color."""
+        return self._light_color
+
+    @light_color.setter
+    def light_color(self, light_color):
+        self._light_color = Color(light_color)
+        self._update_data()
+
+    @property
+    def diffuse_color(self):
+        """The diffuse light color."""
+        return self._diffuse_color
+
+    @diffuse_color.setter
+    def diffuse_color(self, diffuse_color):
+        self._diffuse_color = Color(diffuse_color)
+        self._update_data()
+
+    @property
+    def specular_color(self):
+        """The specular light color."""
+        return self._specular_color
+
+    @specular_color.setter
+    def specular_color(self, specular_color):
+        self._specular_color = Color(specular_color)
+        self._update_data()
+
+    @property
+    def ambient_color(self):
+        """The ambient color."""
+        return self._ambient_color
+
+    @ambient_color.setter
+    def ambient_color(self, color):
+        self._ambient_color = Color(color)
+        self._update_data()
+
+    @property
+    def shininess(self):
+        """The shininess."""
+        return self._shininess
+
+    @shininess.setter
+    def shininess(self, shininess):
+        self._shininess = shininess
+        self._update_data()
+
+    def _update_data(self):
+        if not self._attached:
+            return
+        self.vcode['light_dir'] = self._light_dir
+        self.fcode['shininess'] = self._shininess
+        self.fcode['light_color'] = self._light_color.rgb
+        self.fcode['ambient_color'] = self._ambient_color.rgb
+        self.fcode['diffuse_color'] = self._diffuse_color.rgb
+        self.fcode['specular_color'] = self._specular_color.rgb
+        self.fcode['flat_shading'] = 1 if self._shading == 'flat' else 0
+        self.fcode['shading_enabled'] = 1 if self._shading is not None else 0
+        normals = self._visual().mesh_data.get_vertex_normals(indexed='faces')
+        self._normals.set_data(normals, convert=True)
+
+    def _attach(self, visual):
+        # vertex shader
+        vert_post = visual._get_hook('vert', 'post')
+        vert_post.add(self.vertex_program)
+        render2scene = visual.transforms.get_transform('render', 'scene')
+        visual2scene = visual.transforms.get_transform('visual', 'scene')
+        scene2doc = visual.transforms.get_transform('scene', 'document')
+        doc2scene = visual.transforms.get_transform('document', 'scene')
```

Comment:
This is a big problem when this is used outside of the `SceneCanvas`. The `transforms` system only exists with the `SceneCanvas`. Except now that I'm looking at the code...this is the way it was in the original Visual so...this never work in a regular Canvas? Or am I missing something?

Reply:
I am not sure about this one. I've only worked with the SceneCanvas for meshes.